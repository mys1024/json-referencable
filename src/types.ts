export interface Referencified {
  refs: Record<string, object | unknown[]>
  root: string
}
