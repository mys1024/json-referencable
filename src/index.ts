interface Referencified {
  refs: Record<string, object | unknown[]>
  root: string
}

function isArray(val: unknown): val is unknown[] {
  return Array.isArray(val)
}

function isObject(val: unknown): val is object {
  return typeof val === 'object'
}

function isArrOrObj(val: unknown): val is object | unknown[] {
  return isArray(val) || isObject(val)
}

function isReferencified(val: any): val is Referencified {
  return isObject(val?.refs) && typeof val?.root === 'string'
}

export function referencify(
  data: object | unknown[],
  prefix = '_ref_',
): Referencified {
  const map: Map<object | unknown[], string> = new Map()
  let acc = 0

  const ref = (d: object | unknown[]) => {
    let id = `${prefix}${acc++}`
    map.set(d, id)
    return id
  }

  const dfs = (d: object | unknown[]) => {
    if (!isArrOrObj(d))
      throw new Error('type error')
    for (const key in d) {
      const val = d[key]
      if (!isArrOrObj(val))
        continue
      let id = map.get(val)
      if (id === undefined) {
        id = ref(val)
        dfs(val)
      }
      d[key] = id
    }
  }

  const root = ref(data)
  dfs(data)

  const refs: Record<string, object | unknown[]> = {}
  for (const [val, key] of map.entries()) {
    refs[key] = val
  }

  return { refs, root }
}

export function stringify(
  data: object | unknown[],
  replacer?: (this: any, key: string, value: any) => any,
  space?: string | number,
): string {
  return JSON.stringify(referencify(data), replacer, space)
}

export function parse<T extends object | unknown[]>(
  json: string,
  prefix = '_ref_',
): T {
  const referencified = JSON.parse(json)
  if (!isReferencified(referencified))
    throw new Error('not a referencable JSON string')

  const { refs, root } = referencified
  const data = refs[root]
  if (!isArrOrObj(data))
    throw new Error(`wrong ref id: ${root}`)

  const dfs = (d: object | unknown[]) => {
    for (const key in d) {
      const maybeId = d[key]
      if (typeof maybeId !== 'string' || !maybeId.startsWith(prefix))
        continue
      const val = refs[maybeId]
      if (!isArrOrObj(val))
        throw new Error(`wrong ref id: ${maybeId}`)
      d[key] = val
      dfs(val)
    }
  }

  dfs(data)

  return data as T
}
