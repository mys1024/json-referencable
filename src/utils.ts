import type { Referencified } from './types'

export function isArray(val: unknown): val is unknown[] {
  return Array.isArray(val)
}

export function isObject(val: unknown): val is object {
  return typeof val === 'object'
}

export function isArrOrObj(val: unknown): val is object | unknown[] {
  return isArray(val) || isObject(val)
}

export function isReferencified(val: any): val is Referencified {
  return isObject(val?.refs) && typeof val?.root === 'string'
}

export function shallowClone<T extends object | unknown[]>(data: T): T {
  if (!isArrOrObj(data))
    throw new Error('not an array or object')
  else if (isArray(data))
    return [...data] as T
  else
    return {...data}
}

export function deepClone<T extends object | unknown[]>(data: T): T {
  if (!isArrOrObj(data))
    throw new Error('not an array or object')

  const clones: Map<object | unknown[], object | unknown[]> = new Map()

  const dfs = (d: object | unknown[]) => {
    for (const key in d) {
      const val = d[key]
      if (!isArrOrObj(val))
        continue
      let clone = clones.get(val)
      if (clone === undefined) {
        clone = shallowClone(val)
        clones.set(val, clone)
        dfs(clone)
      }
      d[key] = clone
    }
  }

  const clone = shallowClone(data)
  clones.set(data, clone)
  dfs(clone)

  return clone
}
