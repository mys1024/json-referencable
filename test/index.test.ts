import { describe, expect, it } from 'vitest'
import { referencify, stringify, parse } from '../src/index'

interface Student {
  name: string
  id: number
  class: Class,
  friends: Student[]
}

interface Class {
  students: Student[]
}

function createTestData() {
  const c1: Class = {
    students: []
  }
  const s1: Student = {
    name: 'A',
    id: 1,
    class: c1,
    friends: []
  }
  const s2: Student = {
    name: 'B',
    id: 2,
    class: c1,
    friends: []
  }
  c1.students.push(s1, s2)
  s1.friends.push(s2)
  s2.friends.push(s1)
  return c1
}

describe.concurrent('referencify', () => {
  it('test1', () => {
    expect(referencify(createTestData())).toMatchSnapshot()
  })
})

describe.concurrent('stringify ', () => {
  it('test1', () => {
    expect(stringify(createTestData())).toMatchSnapshot()
  })
})

describe.concurrent('parse ', () => {
  it('test1', () => {
    const json = `{
      "refs": {
        "_ref_0": ["_ref_0"]
      },
      "root": "_ref_0"
    }`
    const arr: unknown[] = []
    arr.push(arr)
    expect(parse(json)).toMatchObject(arr)
  })
})

describe.concurrent('stringify and parse ', () => {
  it('test1', () => {
    const arr: unknown[] = []
    arr.push(arr)
    const parsedArr = parse<unknown[]>(stringify(arr))
    expect(parsedArr).toBe(parsedArr[0])
  })

  it('test2', () => {
    const obj: { self: unknown } = { self: undefined }
    obj.self = obj
    const parsedObj = parse<{ self: unknown }>(stringify(obj))
    expect(parsedObj).toBe(parsedObj.self)
  })

  it('test3', () => {
    const data = createTestData()
    const obj: { self: unknown } = { self: undefined }
    obj.self = obj
    const parsedData = parse<Class>(stringify(data))
    expect(parsedData.students[0].class).toBe(parsedData)
    expect(parsedData.students[0].class).toBe(parsedData.students[1].class)
    expect(parsedData.students[0].friends[0]).toBe(parsedData.students[1])
    expect(parsedData.students[1].friends[0]).toBe(parsedData.students[0])
  })
})
