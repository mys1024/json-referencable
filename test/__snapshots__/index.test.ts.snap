// Vitest Snapshot v1

exports[`referencify > test1 1`] = `
{
  "refs": {
    "_ref_0": {
      "students": "_ref_1",
    },
    "_ref_1": [
      "_ref_2",
      "_ref_4",
    ],
    "_ref_2": {
      "class": "_ref_0",
      "friends": "_ref_3",
      "id": 1,
      "name": "A",
    },
    "_ref_3": [
      "_ref_4",
    ],
    "_ref_4": {
      "class": "_ref_0",
      "friends": "_ref_5",
      "id": 2,
      "name": "B",
    },
    "_ref_5": [
      "_ref_2",
    ],
  },
  "root": "_ref_0",
}
`;

exports[`stringify  > test1 1`] = `"{\\"refs\\":{\\"_ref_0\\":{\\"students\\":\\"_ref_1\\"},\\"_ref_1\\":[\\"_ref_2\\",\\"_ref_4\\"],\\"_ref_2\\":{\\"name\\":\\"A\\",\\"id\\":1,\\"class\\":\\"_ref_0\\",\\"friends\\":\\"_ref_3\\"},\\"_ref_3\\":[\\"_ref_4\\"],\\"_ref_4\\":{\\"name\\":\\"B\\",\\"id\\":2,\\"class\\":\\"_ref_0\\",\\"friends\\":\\"_ref_5\\"},\\"_ref_5\\":[\\"_ref_2\\"]},\\"root\\":\\"_ref_0\\"}"`;
